internal/pkg/service/common/etcdop/mutex.go

```diff
@@ -0,0 +1,347 @@
+package etcdop
+
+import (
+	"context"
+	"sync"
+
+	"go.etcd.io/etcd/api/v3/etcdserverpb"
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"go.etcd.io/etcd/client/v3/concurrency"
+)
+
+// Mutex provides distributed locking, the interface is compatible with the concurrency.Mutex.
+// The difference is that the Mutex also works as a local lock,
+// so if the Lock method is called twice in a goroutine,
+// the second call is blocked until the Unlock is called at first.
+//
+// Implementation:
+//   - Mutex is only a user-facing interface to the internal struct mutexStore.
+//   - The local lock is implemented via a channel, so it is possible to cancel waiting for the lock via the context.
+//   - The "usage" counter prevents memory leaks, unused references are deleted.
+type Mutex struct {
+	store  *mutexStore
+	name   string
+	locked *activeMutex
+}
```

Comment:
Go native lock seem like `mutex.Lock()`, there is no` mutex.Lock(ctx)`, but all our operations have a timeout.

But context aware locks can be simple implemented via a channel:
https://h12.io/article/go-pattern-context-aware-lock